external/xpartition/xpartition.py

```diff
@@ -0,0 +1,402 @@
+):
+    for da in ds.data_vars.values():
+        if isinstance(da.data, dask.array.Array):
+            partition_dims = [dim for dim in dims if dim in da.dims]
+            da.partition.write(store, ranks, partition_dims, rank)
+
+
+class Map(Sequence):
+    """Lazy sequence"""
+
+    def __init__(self, func, seq):
+        self.seq = seq
+        self.func = func
+
+    def __getitem__(self, i):
+        return self.func(self.seq[i])
+
+    def __len__(self):
+        return len(self.seq)
+
+
+@xr.register_dataarray_accessor("partition")
+class PartitionDataArrayAccessor:
+    def __init__(self, xarray_obj):
+        self._obj = xarray_obj
+        if not isinstance(self._obj.data, dask.array.Array):
+            raise ValueError(
+                "The partition accessor is only valid for dask-backed arrays."
+            )
+
+    def _meta_array(self, chunks: Dict[Hashable, int]) -> xr.DataArray:
+        dummy_data = dask.array.zeros(self._obj.blocks.shape)
+        da = xr.DataArray(dummy_data, dims=self._obj.dims, name="blocks")
+        return da.chunk(chunks)
+
+    def _optimal_meta_chunk_sizes(
+        self, ranks: int, dims: Sequence[Hashable]
+    ) -> Dict[Hashable, int]:
+        """Determine the optimal meta chunk sizes for the DataArray.
+
+        Partitions are prioritized based on the ordering of the dims
+        provided.  Priority means we will first make the meta chunk
+        size one along those dimensions before moving to larger meta
+        chunk sizes.
+
+        Parameters
+        ----------
+        ranks : int
+            Total number of ranks available to partition across.
+        dims : Sequence[Hashable]
+            Dimensions to partition among; if a dimension is left out
+            no partitions will be made along that dimension.
+
+        Returns
+        -------
+        Dict[Hashable, int]
+        """
+        chunk_sizes = {}
+        for dim in dims:
+            block_sizes = []
+            for d, s in chunk_sizes.items():
+                block_size = math.ceil(self._obj.blocks.sizes[d] / s)
+                block_sizes.append(block_size)
+            blocks = np.product(block_sizes)
+            size = math.ceil(self._obj.blocks.sizes[dim] / (ranks // blocks))
+            chunk_sizes[dim] = min(size, self._obj.blocks.sizes[dim])
+        return chunk_sizes
+
+    def partition(self, ranks, dims) -> Partition:
+        """Compute a ranks-sized partition respecting dask block boundaries
+
+        Parameters
+        ----------
+        ranks : int
+            Total number of ranks available to partition across.
+        dims : Sequence[Hashable]
+            Dimensions to partition among; if a dimension is left out
+            no partitions will be made along that dimension.
+
+        Returns
+        -------
+        a list of disjoint regions whose union is the full coordinate space
+        """
+        return Map(functools.partial(self._indexers, ranks, dims), list(range(ranks)))
```

Comment:
>  think there is a clear more fundamental strategy for improving the performance of indexers

This would be nice since I could imagine some algos wanting the full partitioning.